Clarify name bindings in namespaces. (#3407)
carbon-language/carbon-lang: proposals/p3407.md

<!-- toc -->

## Table of contents

-   [Abstract](#abstract)
-   [Problem](#problem)
-   [Background](#background)
-   [Proposal](#proposal)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Allow prefixing a tuple binding pattern with a namespace](#allow-prefixing-a-tuple-binding-pattern-with-a-namespace)
    -   [Allow binding patterns to declare names in multiple namespaces](#allow-binding-patterns-to-declare-names-in-multiple-namespaces)
    -   [Allow declaring names in namespaces not owned by the current scope](#allow-declaring-names-in-namespaces-not-owned-by-the-current-scope)
    -   [Disallow declaring namespaces other than the file scope](#disallow-declaring-namespaces-other-than-the-file-scope)

<!-- tocstop -->

## Abstract

-   Require namespace members be declared in the same name scope as the
    namespace is declared.
-   Allow binding patterns to directly declare names in namespaces.
-   Disallow using different namespaces in the same binding pattern.

## Problem

While the trivial case of `var NS.a` seems to be supported as a consequence of
proposal

> [#107: Code and name organization](https://github.com/carbon-language/carbon-lang/pull/107),

Reviewer: I went looking there and didn't see a mention of this case for patterns. Is there a specific part I should be looking at?

Author: Broadly speaking, #107 doesn't get into pattern syntax details. But a big reason namespaces were added was C++ compatibility. C++ allows variables in namespaces, and I'm not aware of any decisions to the contrary, so isn't this supported as a consequence?

Do you want more comments regarding #107 being ambiguous beyond just "seems to be supported"? An alternative evaluating not supporting `var NS.a`? Do you want to disallow `NS.a`?

Reviewer: The `var` syntax is `var PATTERN [...]`, and I couldn't find anywhere in #107 that says that you can use `NS.a: Type` as a pattern. The pattern matching proposal specifically says you can only use an identifier on the left-hand side of a `:`. So my conclusion was that `var NS.a: Type` was not supported by the design prior to this proposal -- to me, this part is an (important!) new syntactic feature, not something that was already supported.

Reading this again, I think this sentence wasn't supposed to be saying something about `var` and patterns, so maybe we could just use a different example:

```suggestion
While the trivial case of `class NS.C` seems to be supported as a consequence of
proposal
[#107: Code and name organization](https://github.com/carbon-language/carbon-lang/pull/107),
```

there are a couple syntactic options when binding multiple names.

Also, there's no clear decision around code such as:

```carbon
namespace NS;
class ClassT {
  // Is this a class member accessed through `NS`, or a file scope member inside
  // `NS`? What is its lifetime?

>   var NS.a = 0;

Reviewer: Should this be something like:
```suggestion
  var NS.a: i32 = 0;
```

}
```

This proposal mainly aims to remove these ambiguities.

## Background

Namespaces are covered in
[code and name organization](/docs/design/code_and_name_organization/#namespaces).
Binding patterns are covered in
[pattern matching](/docs/design/pattern_matching.md#binding-patterns).

There's some discussion of `var` in
[values, variables, and pointers](/docs/design/values.md), but it's specific to
locals. That doesn't address other use cases, such as globals or member
variables.

## Proposal

When used to declare names in binding patterns, as in `var` or `let`, all names
must be in the same namespace. `namespace` members must be declared from within
the same scope that declared the `namespace`.

See the changes to
[code and name organization](/docs/design/code_and_name_organization/#namespaces)
for reference.

## Rationale

-   [Performance-critical software](/docs/project/goals.md#performance-critical-software)

    -   Only allowing binding patterns to declare variables in the same
        namespace allows for a guarantee that the memory can be sequentially
        allocated. This allows for copy elision on initializers with function

>         calls.

Reviewer: I don't understand this. How does the choice of namespace affect the memory layout?

Author: This fell out of sync with the removal of related commentary in the doc. Removed for simplicity.

### Allow declaring names in namespaces not owned by the current scope

We could allow declaring names in namespaces not owned by the current scope. For
example:

```carbon
namespace NS;
class ClassT {
  var NS.val: i32;

  class NS.ChildT {}
}
```

Here, `package.NS.val` would be a global, but `ClassT.NS.val` looks more like an
instance member. It's also unclear whether `ClassT.NS.val` (or
`instance.NS.val`) could be used to reference the produced variable, since `NS`
is not inside `ClassT`'s name scope. The naming problems extend to non-binding
declarations such as `NS.ChildT`, too.

Disallowing using namespaces to cross name scopes is consistent with rules that
generally prevent declaring names in other name scopes, such as:

```carbon
class A {
    class B {
        // `C` must be declared directly inside `A`.
        class A.C;
    }
}

// `D` must be declared within `A`, even if separately defined.
class A.D {}
```

Both the `namespace` declaration and names declared within the `namespace` must
be written in the same name scope. This avoids name lookup ambiguities, and
builds consistency in name scope boundaries across declarations.

> ### Disallow declaring namespaces other than the file scope

Reviewer: I believe disallowing this is the status quo ante, at least by implication. Looking for mentions of "namespace" in docs/design, I didn't find anywhere that directly and clearly states as a rule that namespaces are only allowed at file scope, but I found several places that assume that to be the case ([example](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/code_and_name_organization/README.md#namespaces:~:text=A%20namespace%20declaration%20adds%20the%20first%20identifier%20in%20the%20name%20path%20as%20a%20name%20in%20the%20file%27s%20namespace)).

If the intent of this proposal is to explicitly permit class- and function-scope namespaces, then I think that would be a design change, and I think we need more investigation into consequences and we need rationale. If not, I'd suggest we just drop this section for now (and the `FunctionNS` parts of the example in the other file) and leave it to another proposal to dig into this.

Author: Removing -- or rather, rephrasing for the inverse.

For reference, I'd talked about this with @chandlerc and he'd supported doing it -- leads might want to discuss this. But I'm hoping this sets up the proposal for faster completion.

Author: FYI, unresolving just in case chandlerc is wondering why this change occurred.

We could disallow declaring namespaces in scopes other than the file scope.
Combined with requiring members of namespaces to only be declared at the file
scope, this would prevent their use in other contexts.

This would be consistent with C++. However, we believe the current design
doesn't actively interfere with Carbon, and it could prove useful in some
situations. For example, perhaps a complex class would find a member namespace
useful:

```carbon
class Complex {
   namespace OptionSet1;
   class OptionSet1.MemberClassA;
   class OptionSet1.MemberClassB;

   namespace OptionSet2;
   class OptionSet2.MemberClassC;
   class OptionSet2.MemberClassD;

> }

Reviewer: Our qualified name lookup rules definitely don't do the right thing for class-scope namespaces as they stand. (Eg, we won't do instance binding when naming a field that's a member of a namespace that's a member of a class.) I'm not sure how complicated that is to fix, but it would fit quite badly into the model in explorer for instance, where we want to know at the point where we form a member name whether it's an instance member or not, which we don't know when the member name is a namespace name.

Author: I think there are significant design problems with the explorer, but we're archiving it. How much do you expect it to factor into language design decisions?

With toolchain, my expectation is that it wouldn't be complicated. Though certainly one choice would be to say that `namespace` is equivalent to a type declaration, not a variable declaration, and so cannot be used for instance members. But I think this change moots it.

Reviewer: I think there are language semantics issues here that go beyond the implementation -- if we have `x: Complex`, what is the type of `x.OptionSet1`? If it can have instance members, then is it a class? Or is it some new kind of class-namespace-fragment thing? Or something else? Is it a first-class value that can be passed around, or like a namespace can it only be used in limited ways?

To be clear: I think these questions can be answered. But I think we shouldn't add the functionality without answering them, and I don't think we should block this proposal on figuring out the answers :) I'm very happy for you to be setting this to the side for now.

    -   Only allowing binding patterns to declare variables in the same
        namespace allows for a guarantee that the memory can be sequentially
        allocated. This allows for copy elision on initializers with function
        calls.

-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)

    -   Requiring that declarations of multiple names use `NS.a` syntax is
        consistent with the single variable case.
    -   Requiring namespace members be declared while in the same name scope as
        the namespace itself makes lifetimes clearer.

## Alternatives considered

### Allow prefixing a tuple binding pattern with a namespace

We could use the namespace to prefix the binding tuple. For example:

```carbon
var NS.(a: i32, b: i32) = (3, 4);
```

It's rare that we would have a single statement declare multiple names. As a
consequence, the separation of the namespace qualifier from the declared
identifier might end up unique to this syntax. In that context, we prefer `NS.a`
for consistency with other cases, such as `class NS.class`.

> ### Allow mixing namespaces within a binding pattern

Reviewer 2: This also seems to cover unrelated scopes?

```suggestion
### Allow using namespaces in unrelated scopes or mixing them within a binding pattern
```

Author: I think these are two separate alternatives... [but maybe I don't understand what you're getting at? do you want me to add an alternative "Don't allow using namespaces in unrelated scopes?", i.e. that namespaces can only be used locally?]

Reviewer 2: Sure, but then I'd separate out the rationale too. I think the static member thing is more attached to the unrelated scope nesting than mixing within a binding, although certainly combining the two exacerbates the rationale.

Author: I've split and rephrased the sections. Note I've also dropped the `static` mentions: I think my original intent with that wasn't working for you, and it didn't feel like the rephrasing needed it.

Reviewer 2: I somewhat understood the `static` thing, but I think your new version is really nice and simple, so let's go with it. =D Thanks!!

We could allow mixing namespaces in a binding pattern. For example:

```carbon
var (e: i32, NS.f: i32) = (5, 6);
var (FirstNS.g: i32, SecondNS.h: i32) = (7, 8);
```

> That would offer the possibility of mixing use inside scopes:
> 
> ```carbon
> namespace NS;
> 
> class ClassT {
>   var (static class_member: i32,
>        NS.namespace_member: i32,
>        package.package_member: i32) = InitData();
> }
> ```
> 
> Note here that static syntax for class members is not defined: I'm assuming for
> argument's sake that we expect something similar, but this could also be:
> 
> ```
>   let (class_member: i32,
>        var NS.namespace_member: i32,
>        var package.package_member: i32) = InitData();
> ```
> 
> It's not obvious how memory for the above should be allocated. There are two
> options:
> 
> -   The three values are sequentially allocated, even though they belong to
>     different name scopes. This allows for return value optimizations to work,
>     but may cause issues for how we consider name scopes and lifetimes.
> -   The three values are separately allocated, grouping allocations by name
>     scope. This breaks return value optimizations, but simplifies requirements
>     around name scopes and lifetimes.
> 
> Both options have issues; it would likely prove difficult to set this up such
> that return value optimizations would consistently work, across all potential
> uses of this syntax. That would introduce some variability in performance based
> on unclear context, something we would like to avoid.
> 
> Without mixing namespaces, static initialization case can still be addressed,
> albeit with more work. For example:
> 
> ```carbon
> namespace NS;
> 
> let init_data: auto = InitData();
> 
> class ClassT {
>   var static class_member: i32 = init_data[0];
> }
> 
> var NS.namespace_member: i32 = init_data[1];
> var package_member: i32 = init_data[2];
> ```
> 
> Rather than forcing a choice, this alternative is discarded in order to require
> syntax -- and resulting performance -- be clearer.

Reviewer 2: Could we can have a much simpler/shorter rationale here without fully exploring the static class variable space? Maybe something like:

````suggestion
The mixture of two different prefixes in a single declaration alone seems likely to be confusing, and doesn't have much motivation.

If we allow namespaced names within unrelated scopes, it would also imply the possibility of mixing use inside *class* scopes:

```carbon
namespace NS;

class ClassT {
  var (class_member: i32,
       NS.namespace_member: i32,
       package.package_member: i32);
}
```

Either these are all instance members, which would be surprising in terms of the naming conventions for instance members (for example requiring looking up the namespace name after the object in `x.NS.namespace_member`), or it would require the qualified names to be *static* members (in C++ terminology).

Mixing static and non-static members in a single declaration seems fundamentally problematic due to lifetimes.

We don't have a design for static class members like this yet, and it doesn't seem like we should introduce or motivate them with namespacing within scopes.

For now, we just disallow the mixing namespaces in single declarations for simplicity and to avoid having to resolve these issues. And we disallow namespaced names within unrelated scopes to avoid the name lookup complexity and any potential confusion over C++-style-`static` or other semantics.
````

Maybe could be further simplified or reduced...

Author: I think this has actually gone in a different direction than the original text. The example does not have unrelated scopes: `ClassT` is within `package`.

I'm going to need to spend time cleaning this up, but not right now. Now I think there's probably a misunderstanding of the point the original text was trying to make.

-->

[Pull request](https://github.com/carbon-language/carbon-lang/pull/3407)

<!-- toc -->

## Table of contents

-   [Abstract](#abstract)
-   [Problem](#problem)
-   [Background](#background)
-   [Proposal](#proposal)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Allow prefixing a tuple binding pattern with a namespace](#allow-prefixing-a-tuple-binding-pattern-with-a-namespace)
    -   [Allow binding patterns to declare names in multiple namespaces](#allow-binding-patterns-to-declare-names-in-multiple-namespaces)
    -   [Allow declaring names in namespaces not owned by the current scope](#allow-declaring-names-in-namespaces-not-owned-by-the-current-scope)
    -   [Allow declaring namespaces in scopes other than the file scope](#allow-declaring-namespaces-in-scopes-other-than-the-file-scope)

<!-- tocstop -->

## Abstract

-   Require namespace members be declared in the same name scope as the
    namespace is declared.
    -   Declaring namespaces outside file scope is disallowed, so this means
        only at file scope.
-   Allow binding patterns to directly declare names in namespaces.

> -   Disallow using different namespaces in the same binding pattern.

Reviewer: ```suggestion
-   Disallow declaring names in multiple namespaces in the same pattern.
```
The previous phrasing read to me as if you can't use `N.a: M.T` if `N` and `M` are different namespaces, but you can use `(N.a: N.T, M.b: M.U)` because that's two different binding patterns.

instance member. It's also unclear whether `ClassT.NS.val` (or
`instance.NS.val`) could be used to reference the produced variable, since `NS`
is not inside `ClassT`'s name scope. The naming problems extend to non-binding
declarations such as `NS.ChildT`, too.

Disallowing using namespaces to cross name scopes is consistent with rules that
generally prevent declaring names in other name scopes, such as:

```carbon
class A {
    class B {
        // `C` must be declared directly inside `A`.
        class A.C;
    }
}

// `D` must be declared within `A`, even if separately defined.
class A.D {}
```

Both the `namespace` declaration and names declared within the `namespace` must
be written in the same name scope. This avoids name lookup ambiguities, and
builds consistency in name scope boundaries across declarations.

### Allow declaring namespaces in scopes other than the file scope

We could allow declaring namespaces in scopes other than the file scope. It's
ambiguous what path should have resulted from proposal
[#107: Code and name organization](https://github.com/carbon-language/carbon-lang/pull/107),

> although some p

Reviewer: Incomplete sentence.

Author: Fixed, finished the thought.
